account_invoice_mode_at_shipping/models/stock_picking.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import models
+
+from odoo.addons.queue_job.job import job
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def action_done(self):
+        res = super().action_done()
+        picking_to_invoice = self.filtered(
+            lambda r: r.sale_id.partner_invoice_id.invoicing_mode == "at_shipping"
+            and r.picking_type_code == "outgoing"
+        )
+        for picking in picking_to_invoice:
+            picking.with_delay()._invoicing_at_shipping()
+        return res
```

Comment:
Avoid looping twice and make it more readable and customizable:
```suggestion
        for picking in picking_to_invoice:
            if picking._invoice_at_shipping():
                picking.with_delay()._invoicing_at_shipping()
        return res
        
    def _invoice_at_shipping(self):
        """Check if current record must be invoiced at shipping."""
        return (
            self.picking_type_code == "outgoing" and
            self.sale_id.partner_invoice_id.invoicing_mode == "at_shipping"
        )
```

Reply:
This has been changed